src/Orleans/Runtime/GrainReference.cs

```diff
@@ -620,26 +646,26 @@ public static GrainReference FromKeyString(string key)
                 {
                     genericStr = null;
                 }
-                return FromGrainId(GrainId.FromParsableString(grainIdStr), genericStr);
+                return FromGrainId(GrainId.FromParsableString(grainIdStr), null, genericStr);
             }
             else if (observerIndex >= 0)
             {
                 grainIdStr = trimmed.Substring(grainIdIndex, observerIndex - grainIdIndex).Trim();
                 string observerIdStr = trimmed.Substring(observerIndex + (OBSERVER_ID_STR + "=").Length);
                 GuidId observerId = GuidId.FromParsableString(observerIdStr);
-                return NewObserverGrainReference(GrainId.FromParsableString(grainIdStr), observerId);
+                return NewObserverGrainReference(GrainId.FromParsableString(grainIdStr), observerId, null);
             }
             else if (systemTargetIndex >= 0)
             {
                 grainIdStr = trimmed.Substring(grainIdIndex, systemTargetIndex - grainIdIndex).Trim();
                 string systemTargetStr = trimmed.Substring(systemTargetIndex + (SYSTEM_TARGET_STR + "=").Length);
                 SiloAddress siloAddress = SiloAddress.FromParsableString(systemTargetStr);
-                return FromGrainId(GrainId.FromParsableString(grainIdStr), null, siloAddress);
```

Comment:
I still need to work out the best approach. The simplest is to add a 'GetGrainFromKeyString(string)` method to `IGrainFactory` or a new service and to move this method there.